### ISSUE TITLE
BUG - TLX & Inverter Systems - Output Power does not return to 0 overnight

### ISSUE THREAD

**Issue by muppet3000** [muppet3000/homeassistant-growatt_server_api#14]
**Describe the bug**
See: https://github.com/home-assistant/core/issues/81393

**Growatt Device Type**

Device(s): 
- TLX aka MIN/MIC/MOD
- Inverter

**Home Assistant information**
 - Home Assistant version: N/A
 - Growatt Server API Integration version: N/A

**Additional context**

This bug has been duplicated from the original in the Home Assistant Core repo and will be fixed here before before being merged up-stream.
Everyone that has contributed to the bugs so far has been tagged in the comments so that they can receive updates from this repo.

**Comment by muppet3000** [comment]
Hi All, sorry for the mass-tagging but I wanted you all to know about the new ticket that I've raised in this repo for adding support for the AC Couple type of inverter to the integration.
As you're hopefully all aware I've moved the plugin out of Core and will primarily maintain it here because it's much easier and quicker to develop.
When there are significant releases I will merge the changes up-stream into Home Assistant Core, but that will be much slower.

Now for the tags......
@vingerha @saentist @glenndehaan

**Comment by Saentist** [comment]
My Inverter is **UE** series, tree phase [datasheet](https://cdn.enfsolar.com/Product/pdf/Inverter/5b309e0a282c6.pdf), but have same problem.

**Comment by muppet3000** [comment]
Thanks @Saentist - I've updated the description to reflect this.

**Comment by vingerha** [comment]
OK, I installed the HACS version now as well on my DEV. 
Knowing you have many more items on the list and this is not very important....when do you think to be able to tackle this?

**Comment by vingerha** [comment]
Update: first results are good, it now (!) shows 0 as output power...

**Comment by muppet3000** [comment]
Thanks for the update. I have an idea of how I can get the values to go back to 0 before midnight when there's no sun, but I need to do it by querying all of the data available. 
I don't think it's that hard to do, I just need to go through the list in priorities. If I get some time though hopefully in the next few weeks, maybe by the end of February?

**Comment by muppet3000** [comment]
@vingerha & @Saentist - I think I've worked out how to implement this and I've been playing with it tonight. I need to do some more testing but may have a fix in the next few days.

**Comment by muppet3000** [comment]
I now have a system set up to accurately reproduce this. I also have some code ready to go that can be used to fix it, once I've reproduced it overnight tonight I'll apply my fix and then we should hopefully have a resolution

**Comment by vingerha** [comment]
OK, if you want I can test it before

**Comment by muppet3000** [comment]
> OK, if you want I can test it before

As soon as I have something I'm happy with I'll push it to the 'dev' branch and then you can use that option in the HACS installer to test.

**Comment by muppet3000** [comment]
@vingerha - I have a fix that I'm now testing on my local install, I just need to see it go through the night and the day tomorrow and then I'll push it onto Dev so that you can test it out.

@Saentist - Are you still seeing this issue with the Inverter system? I've taken a look at the values on your system and they all seem to be zero as I'd expect overnight. Can you please send me a screen shot of the graph that you're expecting to drop to zero so that I can investigate it further? I need to know what statistic is incorrect so that I can set it to the correct value based on the time of day etc.

**Comment by Saentist** [comment]
Extremely short answer
![image](https://user-images.githubusercontent.com/16975883/217381508-83313782-060f-4695-97b4-0a4a3c50f49f.png)

funny part
temperature reset at midnight showing 9C, outside is -7.6C.

**Comment by muppet3000** [comment]
That graph has no legend/key, it's just a bunch of lines on a graph.

I need to know what statistics these are in Home Assistant so that I can find out where they're taken from on the API and then implement the correct logic for them.

**Comment by Saentist** [comment]
Sorry
![image](https://user-images.githubusercontent.com/16975883/217390047-6a608295-4dfa-4e18-bba3-f4e2e9c1085e.png)
Voltage and temperature

**Comment by muppet3000** [comment]
> Sorry ![image](https://user-images.githubusercontent.com/16975883/217390047-6a608295-4dfa-4e18-bba3-f4e2e9c1085e.png) Voltage and temperature

Thanks, I'll try and take a look at that in more detail tonight, I'll try and find the equivalent values on the app as well, I reckon the app may be using a different endpoint for those values.

**Comment by muppet3000** [comment]
@vingerha - My test last night was half successful. It worked, but the piece of code I wrote to handle the window between midnight and the chart starting never worked, so I need to find out why that's happening.
I've added more debug code and will test again tonight.

**Comment by muppet3000** [comment]
@Saentist - I didn't get the chance to look at this again today, I've been too busy fighting fires to do with the servers keep being unavailable.
I'll try and take another look at your issues separately.
If I've managed to fix the issue that @vingerha has, I'll re-open your issue as a new bug so I can close the original issue off.

**Comment by Saentist** [comment]
off-line from 15:44 GMT no data in HA ;(

**Comment by muppet3000** [comment]
> off-line from 15:44 GMT no data in HA ;(

Not related to this bug.

**Comment by vingerha** [comment]
@muppet3000 ...if you still like me to test this, do send me a link to the code. I can run it for a couple of days then

**Comment by muppet3000** [comment]
> @muppet3000 ...if you still like me to test this, do send me a link to the code. I can run it for a couple of days then

Well I think I have the fix working now, however the test I was running overnight with the credentials you gave me for testing seem to have been locked like we're seeing over here: https://github.com/muppet3000/homeassistant-growatt_server_api/issues/15

I was only polling every 5 minutes so they shouldn't have been blocked but they seem to have been. I've stopped that system now, but you may need to log into the website to generate a new set of credentials for use with home assistant - sorry.

**Comment by vingerha** [comment]
I did not change them and with me they are still working both app as website allow me to login

**Comment by muppet3000** [comment]
> I did not change them and with me they are still working both app as website allow me to login

Ok that's really weird, I can't explain why they've stopped working for me, unless they're doing something clever to block just my IP address.

Either way, I think I've fixed the bug now, so I'll push it to Dev this evening and you can update the plugin to test it out.

Sorry it's taken so long, with all the problems over the last 2 days with the Growatt servers I haven't been able to give this the attention I wanted to.

**Comment by vingerha** [comment]
No excuses needed...thanks for following up !

**Comment by vingerha** [comment]
So... there is something wrong though... I just noticed that the growatt integration is no longer working ...neither the one on core neither your hacs one

**Comment by Saentist** [comment]
> > off-line from 15:44 GMT no data in HA ;(
> 
> Not related to this bug.

![image](https://user-images.githubusercontent.com/16975883/217918952-7c3961f3-74e0-407d-aaa2-7ba88c647664.png)

**Comment by muppet3000** [comment]
> So... there is something wrong though... I just noticed that the growatt integration is no longer working ...neither the one on core neither your hacs one. The panels connect well as I can see the production moving but not the integrations. I reset the password and reconnected web / app...can still see the panels but the HACS integration refuses to connect with the same credentials
> 
> ```
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.10/site-packages/aiohttp/web_protocol.py", line 435, in _handle_request
>     resp = await request_handler(request)
>   File "/usr/local/lib/python3.10/site-packages/aiohttp/web_app.py", line 504, in _handle
>     resp = await handler(request)
>   File "/usr/local/lib/python3.10/site-packages/aiohttp/web_middlewares.py", line 117, in impl
>     return await handler(request)
>   File "/usr/src/homeassistant/homeassistant/components/http/security_filter.py", line 60, in security_filter_middleware
>     return await handler(request)
>   File "/usr/src/homeassistant/homeassistant/components/http/forwarded.py", line 100, in forwarded_middleware
>     return await handler(request)
>   File "/usr/src/homeassistant/homeassistant/components/http/request_context.py", line 28, in request_context_middleware
>     return await handler(request)
>   File "/usr/src/homeassistant/homeassistant/components/http/ban.py", line 81, in ban_middleware
>     return await handler(request)
>   File "/usr/src/homeassistant/homeassistant/components/http/auth.py", line 236, in auth_middleware
>     return await handler(request)
>   File "/usr/src/homeassistant/homeassistant/components/http/view.py", line 136, in handle
>     result = await result
>   File "/usr/src/homeassistant/homeassistant/components/config/config_entries.py", line 180, in post
>     return await super().post(request, flow_id)
>   File "/usr/src/homeassistant/homeassistant/components/http/data_validator.py", line 73, in wrapper
>     result = await method(view, request, data, *args, **kwargs)
>   File "/usr/src/homeassistant/homeassistant/helpers/data_entry_flow.py", line 110, in post
>     result = await self._flow_mgr.async_configure(flow_id, data)
>   File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 286, in async_configure
>     result = await self._async_handle_step(
>   File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 374, in _async_handle_step
>     result: FlowResult = await getattr(flow, method)(user_input)
>   File "/config/custom_components/growatt_server_api/config_flow.py", line 63, in async_step_user
>     self.user_id = login_response["user"]["id"]
> KeyError: 'user'
> ```

Would you be able to generate me a user using this FAQ: https://github.com/muppet3000/homeassistant-growatt_server_api/blob/dev/FAQ.md#how-do-i-share-my-credentials-with-you-for-testingadding-new-features-for-my-system-type so that I can try and work out how the US server is breaking things? I think they must have changed things in a different way.

**Comment by muppet3000** [comment]
> > > off-line from 15:44 GMT no data in HA ;(
> > 
> > 
> > Not related to this bug.
> 
> ![image](https://user-images.githubusercontent.com/16975883/217918952-7c3961f3-74e0-407d-aaa2-7ba88c647664.png)

Sorry, what I meant was that the issues with connectivity to the servers are being discussed over here: https://github.com/muppet3000/homeassistant-growatt_server_api/issues/15 on the thread that you started, I'm trying to keep the discussions in the same place that's all.

**Comment by muppet3000** [comment]
Closing - fixed in https://github.com/muppet3000/homeassistant-growatt_server_api/releases/tag/v1.0.2

@Saentist I will open another bug for your specific issues and link it from here.

**Comment by vingerha** [comment]
@muppet3000 user : vingerha1   and the same password you already know

**Comment by muppet3000** [comment]
> @muppet3000 user : vingerha1 and the same password you already know

Thanks, I've tested it on my phone and it works fine, I'll trial it with the command line tools I have over the weekend. I have a small idea of what might be happening but have no idea on how much effort it will be to fix. I'll see what I can do though.

**Comment by vingerha** [comment]
Thanks and let's see then. I read the other thread and this does not look very positive on the long run for a 'free' solution. I donot mind paying but it depends...also have Tuya stuff and when they started bugging about licenses I kicked them all out...horribly expensive

**Comment by muppet3000** [comment]
I thought there was a free developer license that you had to renew every 6 months with Tuya? That's what I have set up with my Tuya devices in Home Assistant.

Either way, given that Growatt have no official API support now, I'll be surprised if they add it any time in the future. Ultimately I like the idea of doing things locally and cutting out the unreliable servers from the loop. However, in the meantime I'd like to try and unblock the problem you're having with the server-us endpoint. Leave it with me and I'll see what I can do.

**Comment by Saentist** [comment]
Is "Energy today" need also to show zero before midnight?
![image](https://user-images.githubusercontent.com/16975883/220450506-9e07d1b5-a245-448b-b4bd-68a5f92b5917.png)
